http/middleware/idempotency/inmem.go

```diff
@@ -0,0 +1,78 @@
+package idempotency
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+type InMemStore struct {
+	mu        sync.RWMutex
+	inFlight  map[string][]byte // key -> request signature
+	responses map[string]*StoredResponse
+}
+
+// NewInMemStore initializes an in-memory store.
+func NewInMemStore() *InMemStore {
+	return &InMemStore{
+		inFlight:  make(map[string][]byte),
+		responses: make(map[string]*StoredResponse),
+	}
+}
+
+func (s *InMemStore) InsertInFlight(_ context.Context, key string, requestSignature []byte) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If we already have a marker or a stored response, we should decide
+	if _, ok := s.inFlight[key]; ok {
+		return errors.New("already in-flight") //nolint:err113 // it's mostly for testing
+	}
+
+	if _, ok := s.responses[key]; ok {
+		return errors.New("already completed") //nolint:err113 // it's mostly for testing
+	}
+
+	s.inFlight[key] = requestSignature
+
+	return nil
+}
```

Comment:
I think we can lock the idempotent key. It limits the concurrency to 1.  
The second request will always get the result from the first request.
```
func handler() {
   s.Lock(key)
   defer s.Unlock(key)
    resp := s.GetStoredResponse()
    if (resp == null) {
        resp = process_request(req)
    }
    s.MarkComplete(key, req, resp)  
}
```

Reply:
One drawback of lock and InsertInFlight is we have higher risk to have inconsistent btw lock, inflight, responses cache.